File: src/engine/adapters/redis/read.ts
import type { Redis } from 'ioredis';
import {
  createKeyValuePageBuilder,
  type KeyValuePage,
  type PagePosition,
} from '../../../shared/protocol/page';
import type { OpCtx, ReadRequest } from '../adapter';
import { AdapterError } from '../errors';
import { decodePageToken, encodePageToken, requestFingerprint } from '../sql-text';
import { mapError } from './errors';

// Never an unbudgeted SCAN (ground rules) — same per-round-trip COUNT hint as catalog.ts.
const SCAN_COUNT = 1000;
const LIST_WINDOW = 500; // LRANGE window cap, mirrors MAX_PAGE_SIZE discipline

interface KeyMeta {
  redisType: KeyValuePage['redisType'] | 'none';
  ttlMs: number | null;
  memoryBytes: number | null;
}

const KNOWN_TYPES = new Set(['string', 'hash', 'list', 'set', 'zset', 'stream']);

async function readMeta(conn: Redis, key: string, ctx: OpCtx): Promise<KeyMeta> {
  let rawType: string;
  try {
    rawType = await conn.type(key);
  } catch (err) {
    throw mapError(err);
  }
  // A key gone at read time is an ordinary query-time condition (expired/deleted concurrently),
  // not a connection failure — E_QUERY, deliberately not E_NOT_FOUND (P9's D10).
  if (rawType === 'none') throw new AdapterError('E_QUERY', `key no longer exists: ${key}`);
  if (!KNOWN_TYPES.has(rawType)) {
    throw new AdapterError('E_UNSUPPORTED', `unsupported redis type for ${key}: ${rawType}`);
  }
  let pttl: number;
  try {
    pttl = await conn.pttl(key);
  } catch (err) {
    throw mapError(err);
  }
  let memoryBytes: number | null;
  try {
    memoryBytes = await conn.memory('USAGE', key);
  } catch {
    memoryBytes = null; // best-effort (§8.8)
  }
  if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');
  return {
    redisType: rawType as KeyValuePage['redisType'],
    ttlMs: pttl >= 0 ? pttl : null,
    memoryBytes,
  };
}

async function readString(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  ctx: OpCtx,
): Promise<KeyValuePage> {
  if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');
  let value: string | null;
  try {
    value = await conn.get(key);
  } catch (err) {
    throw mapError(err);
  }
  const builder = createKeyValuePageBuilder({
    redisType: 'string',
    ttlMs: meta.ttlMs,
    memoryBytes: meta.memoryBytes,
  });
  if (value !== null) builder.push('value', value);
  const position: PagePosition = {
    offset: 0,
    pageSize: 1,
    hasMore: false,
    nextToken: null,
    prevToken: null,
    strategy: 'offset',
  };
  return builder.finish(position);
}

// Shared cursor-loop body for hash/set/zset (§8.8's per-type renderers): accumulates whole SCAN
// rounds without slicing mid-round, so a round's remaining elements are never dropped (P9's read
// notes) — the page can slightly overshoot req.pageSize, which is fine for a browse-only view.
async function readScanFamily(
  scanOnce: (cursor: string) => Promise<[string, string[]]>,
  pairSize: 1 | 2,
  redisType: KeyValuePage['redisType'],
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
  fingerprint: string,
): Promise<KeyValuePage> {
  if (req.cursor.mode === 'before') {
    throw new AdapterError(
      'E_UNSUPPORTED',
      'redis cursor pagination is forward-only; there is no previous page',
    );
  }
  let cursor = '0';
  if (req.cursor.mode === 'after') {
    [cursor] = decodePageToken(req.cursor.token, fingerprint);
  }

  const builder = createKeyValuePageBuilder({
    redisType,
    ttlMs: meta.ttlMs,
    memoryBytes: meta.memoryBytes,
  });
  let rowCount = 0;
  let exhausted = false;

  do {
    if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');
    const [nextCursor, elements] = await scanOnce(cursor);
    cursor = nextCursor;
    for (let i = 0; i < elements.length; i += pairSize) {
      if (pairSize === 2) builder.push(elements[i], elements[i + 1]);
      else builder.push(String(rowCount), elements[i]);
      rowCount++;
    }
    if (cursor === '0') exhausted = true;
  } while (rowCount < req.pageSize && !exhausted);

  const hasMore = !exhausted;
  const position: PagePosition = {
    offset: null,
    pageSize: req.pageSize,
    hasMore,
    nextToken: hasMore ? encodePageToken([cursor], fingerprint) : null,
    prevToken: null,
    strategy: 'cursor',
  };
  return builder.finish(position);
}

async function readHash(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
  fingerprint: string,
): Promise<KeyValuePage> {
  return readScanFamily(
    async (cursor) => {
      try {
        return await conn.hscan(key, cursor, 'COUNT', SCAN_COUNT);
      } catch (err) {
        throw mapError(err);
      }
    },
    2,
    'hash',
    meta,
    req,
    ctx,
    fingerprint,
  );
}

async function readSet(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
  fingerprint: string,
): Promise<KeyValuePage> {
  return readScanFamily(
    async (cursor) => {
      try {
        return await conn.sscan(key, cursor, 'COUNT', SCAN_COUNT);
      } catch (err) {
        throw mapError(err);
      }
    },
    1,
    'set',
    meta,
    req,
    ctx,
    fingerprint,
  );
}

async function readZSet(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
  fingerprint: string,
): Promise<KeyValuePage> {
  return readScanFamily(
    async (cursor) => {
      try {
        return await conn.zscan(key, cursor, 'COUNT', SCAN_COUNT);
      } catch (err) {
        throw mapError(err);
      }
    },
    2,
    'zset',
    meta,
    req,
    ctx,
    fingerprint,
  );
}

async function readList(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
): Promise<KeyValuePage> {
  if (req.cursor.mode !== 'offset') {
    throw new AdapterError('E_UNSUPPORTED', 'redis list pagination only supports offset paging');
  }
  const offset = req.cursor.offset;
  const limit = Math.min(req.pageSize, LIST_WINDOW);
  let elements: string[];
  try {
    elements = await conn.lrange(key, offset, offset + limit - 1);
  } catch (err) {
    throw mapError(err);
  }
  if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');

  let total: number;
  try {
    total = await conn.llen(key);
  } catch (err) {
    throw mapError(err);
  }

  const builder = createKeyValuePageBuilder({
    redisType: 'list',
    ttlMs: meta.ttlMs,
    memoryBytes: meta.memoryBytes,
  });
  elements.forEach((value, i) => {
    builder.push(String(offset + i), value);
  });

  const hasMore = offset + elements.length < total;
  const position: PagePosition = {
    offset,
    pageSize: req.pageSize,
    hasMore,
    nextToken: null,
    prevToken: null,
    strategy: 'offset',
  };
  return builder.finish(position);
}

async function readStream(
  conn: Redis,
  key: string,
  meta: KeyMeta,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
  fingerprint: string,
): Promise<KeyValuePage> {
  if (req.cursor.mode === 'before') {
    throw new AdapterError(
      'E_UNSUPPORTED',
      'redis stream pagination is forward-only; there is no previous page',
    );
  }
  let startId = '-';
  if (req.cursor.mode === 'after') {
    const [afterId] = decodePageToken(req.cursor.token, fingerprint);
    startId = `(${afterId}`; // exclusive lower bound, per XRANGE syntax
  }

  const limit = req.pageSize + 1; // D24's +1 probe, mirroring the SQL adapters
  let entries: [id: string, fields: string[]][];
  try {
    entries = await conn.xrange(key, startId, '+', 'COUNT', limit);
  } catch (err) {
    throw mapError(err);
  }
  if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');

  const probedExtra = entries.length > req.pageSize;
  const kept = probedExtra ? entries.slice(0, req.pageSize) : entries;

  const builder = createKeyValuePageBuilder({
    redisType: 'stream',
    ttlMs: meta.ttlMs,
    memoryBytes: meta.memoryBytes,
  });
  for (const [id, fields] of kept) {
    const pairs: Record<string, string> = {};
    for (let i = 0; i < fields.length; i += 2) pairs[fields[i]] = fields[i + 1];
    builder.push(id, JSON.stringify(pairs));
  }

  const hasMore = probedExtra;
  const position: PagePosition = {
    offset: null,
    pageSize: req.pageSize,
    hasMore,
    nextToken: hasMore ? encodePageToken([kept[kept.length - 1][0]], fingerprint) : null,
    prevToken: null,
    strategy: 'cursor',
  };
  return builder.finish(position);
}

// D6: per-type dispatch after TYPE + PTTL + best-effort MEMORY USAGE classification.
export async function readKey(
  conn: Redis,
  key: string,
  req: Omit<ReadRequest, 'path'>,
  ctx: OpCtx,
): Promise<KeyValuePage> {
  const meta = await readMeta(conn, key, ctx);
  ctx.setCommand(`TYPE ${key}`);
  const fingerprint = requestFingerprint({
    key,
    pageSize: req.pageSize,
    redisType: meta.redisType,
  });

  switch (meta.redisType) {
    case 'string':
      return readString(conn, key, meta, ctx);
    case 'hash':
      return readHash(conn, key, meta, req, ctx, fingerprint);
    case 'set':
      return readSet(conn, key, meta, req, ctx, fingerprint);
    case 'zset':
      return readZSet(conn, key, meta, req, ctx, fingerprint);
    case 'list':
      return readList(conn, key, meta, req, ctx);
    case 'stream':
      return readStream(conn, key, meta, req, ctx, fingerprint);
    default:
      throw new AdapterError(
        'E_UNSUPPORTED',
        `unsupported redis type for ${key}: ${meta.redisType}`,
      );
  }
}

// D6: exact via O(1) type-length commands — exactCount: true for a single key's count.
export async function countKey(
  conn: Redis,
  key: string,
  ctx: OpCtx,
): Promise<{ value: number; exact: boolean }> {
  let rawType: string;
  try {
    rawType = await conn.type(key);
  } catch (err) {
    throw mapError(err);
  }
  if (rawType === 'none') throw new AdapterError('E_QUERY', `key no longer exists: ${key}`);
  if (ctx.signal.aborted) throw new AdapterError('E_CANCELLED', 'operation was cancelled');

  try {
    switch (rawType) {
      case 'string':
        return { value: 1, exact: true };
      case 'hash':
        return { value: await conn.hlen(key), exact: true };
      case 'set':
        return { value: await conn.scard(key), exact: true };
      case 'zset':
        return { value: await conn.zcard(key), exact: true };
      case 'list':
        return { value: await conn.llen(key), exact: true };
      case 'stream':
        return { value: await conn.xlen(key), exact: true };
      default:
        throw new AdapterError('E_UNSUPPORTED', `unsupported redis type for ${key}: ${rawType}`);
    }
  } catch (err) {
    throw mapError(err);
  }
}

File: shell/internal/adapters/redis/read.go
package redis

import (
	"context"
	"encoding/json"
	"errors"
	"strconv"

	goredis "github.com/redis/go-redis/v9"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/page"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// Never an unbudgeted SCAN (ground rules) — same per-round-trip COUNT hint as catalog.go.
const readScanCount = 1000

var knownTypes = map[string]bool{"string": true, "hash": true, "list": true, "set": true, "zset": true, "stream": true}

type keyMeta struct {
	redisType   string // one of knownTypes, or "" if none applies (checked before use)
	ttlMs       *int64
	memoryBytes *int64
}

// readMeta ports read.ts's readMeta.
func readMeta(ctx context.Context, conn *goredis.Client, key string) (keyMeta, error) {
	rawType, err := conn.Type(ctx, key).Result()
	if err != nil {
		return keyMeta{}, mapError(err)
	}
	// A key gone at read time is an ordinary query-time condition (expired/deleted concurrently),
	// not a connection failure — E_QUERY, deliberately not E_NOT_FOUND (P9's D10).
	if rawType == "none" {
		return keyMeta{}, adapters.New(adapters.CodeQuery, "key no longer exists: "+key, nil)
	}
	if !knownTypes[rawType] {
		return keyMeta{}, adapters.New(adapters.CodeUnsupported, "unsupported redis type for "+key+": "+rawType, nil)
	}
	pttl, err := conn.PTTL(ctx, key).Result()
	if err != nil {
		return keyMeta{}, mapError(err)
	}
	var ttlMs *int64
	if ms := pttl.Milliseconds(); ms >= 0 {
		ttlMs = &ms
	}
	var memoryBytes *int64
	if usage, err := conn.MemoryUsage(ctx, key).Result(); err == nil {
		memoryBytes = &usage
	} // best-effort (§8.8)
	return keyMeta{redisType: rawType, ttlMs: ttlMs, memoryBytes: memoryBytes}, nil
}

func readString(ctx context.Context, conn *goredis.Client, key string, meta keyMeta) (page.KeyValuePage, error) {
	value, err := conn.Get(ctx, key).Result()
	if err != nil && !errors.Is(err, goredis.Nil) {
		return page.KeyValuePage{}, mapError(err)
	}
	builder := page.NewKeyValuePageBuilder("string", meta.ttlMs, meta.memoryBytes, false)
	if !errors.Is(err, goredis.Nil) {
		builder.Push("value", value)
	}
	return builder.Finish(page.UnpagedPosition(1)), nil
}

// scanRoundFn runs one SCAN-family round: cursor in, (elements, nextCursor) out. pairSize is 1
// (set) or 2 (hash/zset) — elements alternate field/value for pairSize 2.
type scanRoundFn func(ctx context.Context, cursor uint64) (elements []string, nextCursor uint64, err error)

// readScanFamily is read.ts's own shared cursor-loop body for hash/set/zset (§8.8's per-type
// renderers): accumulates whole SCAN rounds without slicing mid-round, so a round's remaining
// elements are never dropped — the page can overshoot req.pageSize by up to one round. An offset
// cursor on a cursor-paged key silently restarts the scan from 0 rather than seeking or erroring
// (redis 24, views/keyvalue/state.ts's D40 depends on this).
func readScanFamily(ctx context.Context, scanOnce scanRoundFn, pairSize int, redisType string, meta keyMeta, req readReq, fingerprint string) (page.KeyValuePage, error) {
	if req.Cursor.Mode == "before" {
		return page.KeyValuePage{}, adapters.New(adapters.CodeUnsupported,
			"redis cursor pagination is forward-only; there is no previous page", nil)
	}
	var cursor uint64
	if req.Cursor.Mode == "after" {
		keyValues, err := adapters.DecodePageToken(req.Cursor.Token, fingerprint)
		if err != nil {
			return page.KeyValuePage{}, err
		}
		if len(keyValues) != 1 {
			return page.KeyValuePage{}, adapters.New(adapters.CodeQuery, "malformed page token", nil)
		}
		parsed, err := strconv.ParseUint(keyValues[0], 10, 64)
		if err != nil {
			return page.KeyValuePage{}, adapters.New(adapters.CodeQuery, "malformed page token", nil)
		}
		cursor = parsed
	}

	builder := page.NewKeyValuePageBuilder(redisType, meta.ttlMs, meta.memoryBytes, false)
	rowCount := 0
	exhausted := false

	for {
		if err := adapters.CheckCancelled(ctx); err != nil {
			return page.KeyValuePage{}, err
		}
		elements, nextCursor, err := scanOnce(ctx, cursor)
		if err != nil {
			return page.KeyValuePage{}, mapError(err)
		}
		cursor = nextCursor
		for i := 0; i < len(elements); i += pairSize {
			if pairSize == 2 {
				builder.Push(elements[i], elements[i+1])
			} else {
				builder.Push(strconv.Itoa(rowCount), elements[i])
			}
			rowCount++
		}
		if cursor == 0 {
			exhausted = true
		}
		if rowCount >= req.PageSize || exhausted {
			break
		}
	}

	hasMore := !exhausted
	var nextToken *string
	if hasMore {
		token := adapters.EncodePageToken([]string{strconv.FormatUint(cursor, 10)}, fingerprint)
		nextToken = &token
	}
	position := page.PagePosition{
		Offset: nil, PageSize: req.PageSize, HasMore: hasMore,
		NextToken: nextToken, PrevToken: nil, Strategy: "cursor",
	}
	return builder.Finish(position), nil
}

func readHash(ctx context.Context, conn *goredis.Client, key string, meta keyMeta, req readReq, fingerprint string) (page.KeyValuePage, error) {
	return readScanFamily(ctx, func(ctx context.Context, cursor uint64) ([]string, uint64, error) {
		return conn.HScan(ctx, key, cursor, "", readScanCount).Result()
	}, 2, "hash", meta, req, fingerprint)
}

func readSet(ctx context.Context, conn *goredis.Client, key string, meta keyMeta, req readReq, fingerprint string) (page.KeyValuePage, error) {
	return readScanFamily(ctx, func(ctx context.Context, cursor uint64) ([]string, uint64, error) {
		return conn.SScan(ctx, key, cursor, "", readScanCount).Result()
	}, 1, "set", meta, req, fingerprint)
}

func readZSet(ctx context.Context, conn *goredis.Client, key string, meta keyMeta, req readReq, fingerprint string) (page.KeyValuePage, error) {
	return readScanFamily(ctx, func(ctx context.Context, cursor uint64) ([]string, uint64, error) {
		return conn.ZScan(ctx, key, cursor, "", readScanCount).Result()
	}, 2, "zset", meta, req, fingerprint)
}

// readList is read.ts's readList: offset-only, LRANGE offset..offset+pageSize-1 honouring the
// requested page size in full (P43 iter2 D25/F18 — no LIST_WINDOW clamp).
func readList(ctx context.Context, conn *goredis.Client, key string, meta keyMeta, req readReq) (page.KeyValuePage, error) {
	if req.Cursor.Mode != "offset" {
		return page.KeyValuePage{}, adapters.New(adapters.CodeUnsupported, "redis list pagination only supports offset paging", nil)
	}
	offset := req.Cursor.Offset
	elements, err := conn.LRange(ctx, key, int64(offset), int64(offset+req.PageSize-1)).Result()
	if err != nil {
		return page.KeyValuePage{}, mapError(err)
	}
	if err := adapters.CheckCancelled(ctx); err != nil {
		return page.KeyValuePage{}, err
	}
	total, err := conn.LLen(ctx, key).Result()
	if err != nil {
		return page.KeyValuePage{}, mapError(err)
	}

	builder := page.NewKeyValuePageBuilder("list", meta.ttlMs, meta.memoryBytes, false)
	for i, value := range elements {
		builder.Push(strconv.Itoa(offset+i), value)
	}

	hasMore := int64(offset+len(elements)) < total
	off := offset
	position := page.PagePosition{
		Offset: &off, PageSize: req.PageSize, HasMore: hasMore,
		NextToken: nil, PrevToken: nil, Strategy: "offset",
	}
	return builder.Finish(position), nil
}

// readStream is read.ts's readStream: XRANGE key <start> + COUNT pageSize+1.
func readStream(ctx context.Context, conn *goredis.Client, key string, meta keyMeta, req readReq, fingerprint string) (page.KeyValuePage, error) {
	if req.Cursor.Mode == "before" {
		return page.KeyValuePage{}, adapters.New(adapters.CodeUnsupported,
			"redis stream pagination is forward-only; there is no previous page", nil)
	}
	startID := "-"
	if req.Cursor.Mode == "after" {
		keyValues, err := adapters.DecodePageToken(req.Cursor.Token, fingerprint)
		if err != nil {
			return page.KeyValuePage{}, err
		}
		if len(keyValues) != 1 {
			return page.KeyValuePage{}, adapters.New(adapters.CodeQuery, "malformed page token", nil)
		}
		startID = "(" + keyValues[0] // exclusive lower bound, per XRANGE syntax
	}

	limit := req.PageSize + 1 // D24's +1 probe, mirroring the SQL adapters
	entries, err := conn.XRangeN(ctx, key, startID, "+", int64(limit)).Result()
	if err != nil {
		return page.KeyValuePage{}, mapError(err)
	}
	if err := adapters.CheckCancelled(ctx); err != nil {
		return page.KeyValuePage{}, err
	}

	probedExtra := len(entries) > req.PageSize
	kept := entries
	if probedExtra {
		kept = entries[:req.PageSize]
	}

	builder := page.NewKeyValuePageBuilder("stream", meta.ttlMs, meta.memoryBytes, false)
	for _, entry := range kept {
		pairs := make(map[string]string, len(entry.Values))
		for field, value := range entry.Values {
			if s, ok := value.(string); ok {
				pairs[field] = s
			}
		}
		body, err := json.Marshal(pairs)
		if err != nil {
			return page.KeyValuePage{}, err
		}
		builder.Push(entry.ID, string(body))
	}

	hasMore := probedExtra
	var nextToken *string
	if hasMore {
		token := adapters.EncodePageToken([]string{kept[len(kept)-1].ID}, fingerprint)
		nextToken = &token
	}
	position := page.PagePosition{
		Offset: nil, PageSize: req.PageSize, HasMore: hasMore,
		NextToken: nextToken, PrevToken: nil, Strategy: "cursor",
	}
	return builder.Finish(position), nil
}

// readReq is the field subset of adapters.ReadRequest readKey needs, minus Path.
type readReq struct {
	PageSize int
	Cursor   model.PageCursor
}

// readKey is read.ts's readKey (D6): per-type dispatch after TYPE + PTTL + best-effort MEMORY
// USAGE classification.
func readKey(ctx context.Context, conn *goredis.Client, key string, req readReq, op *adapters.OpCtx) (page.KeyValuePage, error) {
	meta, err := readMeta(ctx, conn, key)
	if err != nil {
		return page.KeyValuePage{}, err
	}
	op.SetCommand("TYPE " + key)
	fingerprint := adapters.RequestFingerprint(struct {
		Key       string `json:"key"`
		PageSize  int    `json:"pageSize"`
		RedisType string `json:"redisType"`
	}{key, req.PageSize, meta.redisType})

	switch meta.redisType {
	case "string":
		return readString(ctx, conn, key, meta)
	case "hash":
		return readHash(ctx, conn, key, meta, req, fingerprint)
	case "set":
		return readSet(ctx, conn, key, meta, req, fingerprint)
	case "zset":
		return readZSet(ctx, conn, key, meta, req, fingerprint)
	case "list":
		return readList(ctx, conn, key, meta, req)
	case "stream":
		return readStream(ctx, conn, key, meta, req, fingerprint)
	default:
		return page.KeyValuePage{}, adapters.New(adapters.CodeUnsupported, "unsupported redis type for "+key+": "+meta.redisType, nil)
	}
}

// countKey is read.ts's countKey (D6): exact via O(1) type-length commands.
func countKey(ctx context.Context, conn *goredis.Client, key string) (adapters.CountResult, error) {
	rawType, err := conn.Type(ctx, key).Result()
	if err != nil {
		return adapters.CountResult{}, mapError(err)
	}
	if rawType == "none" {
		return adapters.CountResult{}, adapters.New(adapters.CodeQuery, "key no longer exists: "+key, nil)
	}
	if err := adapters.CheckCancelled(ctx); err != nil {
		return adapters.CountResult{}, err
	}

	var value int64
	switch rawType {
	case "string":
		value = 1
	case "hash":
		value, err = conn.HLen(ctx, key).Result()
	case "set":
		value, err = conn.SCard(ctx, key).Result()
	case "zset":
		value, err = conn.ZCard(ctx, key).Result()
	case "list":
		value, err = conn.LLen(ctx, key).Result()
	case "stream":
		value, err = conn.XLen(ctx, key).Result()
	default:
		return adapters.CountResult{}, adapters.New(adapters.CodeUnsupported, "unsupported redis type for "+key+": "+rawType, nil)
	}
	if err != nil {
		return adapters.CountResult{}, mapError(err)
	}
	return adapters.CountResult{Value: value, Exact: true}, nil
}
